flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java

```diff
@@ -40,26 +40,33 @@
 
 	private final SettablePositionMarker positionMarker = new SettablePositionMarker();
 
-	private boolean bufferConsumerCreated = false;
-
 	public BufferBuilder(MemorySegment memorySegment, BufferRecycler recycler) {
 		this.memorySegment = checkNotNull(memorySegment);
 		this.recycler = checkNotNull(recycler);
 	}
 
 	/**
-	 * @return created matching instance of {@link BufferConsumer} to this {@link BufferBuilder}. There can exist only
-	 * one {@link BufferConsumer} per each {@link BufferBuilder} and vice versa.
+	 * @return created matching instance of {@link BufferConsumer} to this {@link BufferBuilder}.
 	 */
 	public BufferConsumer createBufferConsumer() {
-		checkState(!bufferConsumerCreated, "There can not exists two BufferConsumer for one BufferBuilder");
-		bufferConsumerCreated = true;
 		return new BufferConsumer(
 			memorySegment,
 			recycler,
 			positionMarker);
 	}
 
+	/**
+	 * @return created matching instance of {@link BufferConsumer} similar with {@link #createBufferConsumer()},
+	 * except for initializing its reader position based on {@link BufferBuilder}'s current writable position .
+	 */
+	public BufferConsumer createPositionedBufferConsumer() {
+		return new BufferConsumer(
+			memorySegment,
+			recycler,
+			positionMarker,
+			positionMarker.position);
```

Comment:
I think you could use `positionMarker.cachedPosition` here. `position` is marked `volatile` for the `BufferConsumer`'s thread. Here we are in the realm of `BufferBuilder`'s thread.